bundler/lib/bundler/definition.rb

```diff
@@ -716,6 +716,10 @@ def converge_specs(specs)
           if s.source.include?(effective_source)
             deps << dep
           else
+            if effective_source.dup.specs.search_all(s.name).empty?
```

Comment:
\nit the check can probably be moved into a separate method so it doesn't expose as many internals.

Reply:
Yes, I will do that, thanks! :+1:

Reply:
I wasn't fully satisfied with checking this at this point, so I'm working on a simpler solution. I marked this a WIP for now.